pkg/util/buffer.go

```diff
@@ -0,0 +1,236 @@
+// see the same data.
+//
+// To read from the buffer, call NewStream(). This returns a channel that
+// streams a copy of the buffer's contents. If there is already data
+// accumulated in the buffer, the channel will begin to read it immediately,
+// then continue to receive any new data written to the buffer until the buffer
+// is closed. Each channel returned by NewStream() receives a copy of the
+// buffer's full contents, regardless of when it was created.
+//
+// StreamBuffer implements io.WriteCloser. When Close() is called, all stream
+// channels are closed and any subsequent writes to the buffer return
+// io.ErrClosedPipe. After the buffer is closed, calls to NewStream return a
+// channel that will receive the full contents of the buffer, then be closed.
+//
+// Care should be taken to ensure that the buffer is always closed when no
+// more writes are expected, because the stream channels are expected to be
+// read from until they are closed. Leaving the buffer open will cause all
+// readers to block indefinitely.
+type StreamBuffer struct {
+	chunksMu sync.RWMutex
+	chunks   *list.List
+	closed   bool
+
+	// A list of channels that the buffer will attempt to write to whenever
+	// new data is written to the buffer. This is used to notify stream readers
+	// that are reading in real time and are awaiting new data.
+	notifiers map[int64]chan<- struct{}
+	nextID    int64 // notifier ID counter
+
+	// guards reads and writes to the current chunk
+	rtMu sync.RWMutex
+}
+
+const maxChunkSize = 4096
+
+var _ io.WriteCloser = (*StreamBuffer)(nil)
+
+func NewStreamBuffer() *StreamBuffer {
+	buf := &StreamBuffer{
+		chunks:    list.New(),
+		notifiers: make(map[int64]chan<- struct{}),
+	}
+	buf.chunks.PushBack(newChunk())
+	return buf
+}
+
+func (b *StreamBuffer) Write(p []byte) (n int, err error) {
+	b.chunksMu.Lock()
+	defer b.chunksMu.Unlock()
+
+	if b.closed {
+		return 0, io.ErrClosedPipe
+	}
+	lenP := len(p)
+
+	// if necessary, split p across multiple chunks
+	for len(p) > 0 {
+		tail := b.acquireLastChunkLocked()
+		chunk := tail.Value.(*chunk)
+		b.rtMu.Lock()
+		// check if p would fit in the current chunk without overflowing
+		remainingSpace := cap(chunk.buf) - len(chunk.buf)
+		if len(p) <= remainingSpace {
+			chunk.buf = append(chunk.buf, p...)
+			p = nil
+		} else {
+			chunk.buf = append(chunk.buf, p[:remainingSpace]...)
+			p = p[remainingSpace:]
+		}
+		b.rtMu.Unlock()
+		for _, nc := range b.notifiers {
+			select {
+			case nc <- struct{}{}:
+			default:
+			}
+		}
+	}
+
+	return lenP, nil
+}
+
+func (b *StreamBuffer) Close() error {
+	b.chunksMu.Lock()
+	defer b.chunksMu.Unlock()
+	if b.closed {
+		return nil
+	}
+	b.closed = true
+	close(b.chunks.Back().Value.(*chunk).sealed)
+	return nil
+}
+
+func (b *StreamBuffer) NewStream(ctx context.Context) <-chan []byte {
+	b.chunksMu.Lock()
+	defer b.chunksMu.Unlock()
+
+	rc := make(chan []byte, 1)
+
+	notifier := make(chan struct{})
```

Comment:
The notification channel needs a buffer or you're liable to miss on notifications that happen after you've stopped advancing in the list and before you start selecting on it.
```suggestion
	notifier := make(chan struct{}, 1)
```

Reply:
Good catch, thanks :+1: